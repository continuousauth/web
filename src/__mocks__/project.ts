import { FullProject } from '../common/types';

export const mockProject = (): FullProject => ({
  secret: 'my_secret',
  enabled: true,
  requester_circleCI: null,
  requester_gitHub: null,
  responder_slack: null,
  id: '123',
  repoName: 'my-repo',
  repoOwner: 'my-owner',
  defaultBranch: 'main',
});
