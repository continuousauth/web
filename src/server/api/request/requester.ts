import axios from 'axios';
import * as debug from 'debug';
import * as express from 'express';
import * as Joi from 'joi';
import { Issuer } from 'openid-client';
import * as jwkToPem from 'jwk-to-pem';
import * as jwt from 'jsonwebtoken';

import { createA } from '../../helpers/a';
import { getGitHubAppInstallationToken } from '../../helpers/auth';
import { validate } from '../../helpers/_middleware';
import { Project, OTPRequest } from '../../db/models';
import { getResponderFor } from '../../responders';
import { Requester } from '../../requesters/Requester';
import { projectIsMissingConfig } from '../../../common/types';

const d = debug('cfa:api:request:requester');
const a = createA(d);

const getRequestReadyProject = (projectId: string) =>
  Project.findOne({
    where: {
      id: projectId,
      enabled: true,
    },
    include: Project.allIncludes,
  });

const getFullRequest = async (projectId: string, requestId: string) => {
  return OTPRequest.findOne({
    where: {
      id: requestId,
      projectId,
    },
    include: [
      {
        model: Project,
        include: Project.allIncludes,
      },
    ],
  });
};

export function createRequesterRoutes<R, M>(requester: Requester<R, M>) {
  const router = express();

  // Middleware to verify the requst is authorized
  router.use(
    `/:projectId/${requester.slug}`,
    validate(
      {
        a,
        params: {
          projectId: Joi.number()
            .integer()
            .required(),
        },
      },
      async (req, res, next) => {
        if ((global as any).__bypassRequesterAuthentication) return next();
        const authHeader = req.header('Authorization');
        if (!authHeader)
          return res.status(403).json({
            error: 'Missing Authorization header',
          });

        const project = await Project.findOne({
          where: {
            id: req.params.projectId,
            enabled: true,
          },
          include: Project.allIncludes,
        });

        const badAuthHeader =
          !project ||
          !authHeader.toLowerCase().startsWith('bearer ') ||
          authHeader.substr('bearer '.length) !== project.secret;
        if (!project || badAuthHeader)
          return res.status(404).json({
            error: 'Project not found, disabled or you are not authorizad to access it',
          });

        if (projectIsMissingConfig(project))
          return res.status(400).json({
            error: 'Project not completely configured',
          });

        next();
      },
    ),
  );

  router.post(`/:projectId/${requester.slug}/test`, (req, res) => res.json({ ok: true }));

  router.post(
    `/:projectId/${requester.slug}/credentials`,
    validate(
      {
        a,
        params: {
          projectId: Joi.number()
            .integer()
            .required(),
        },
        body: {
          token: Joi.string().required(),
        },
      },
      async (req, res) => {
        const project = await getRequestReadyProject(req.params.projectId);
        if (!project) return res.status(404).json({ error: 'Could not find project' });

        const config = requester.getConfigForProject(project);
        if (!config)
          return res.status(422).json({ error: 'Project is not configured to use this requester' });

        const disoveryUrl = await requester.getOpenIDConnectDiscoveryURL(project, config);
        if (!disoveryUrl)
          return res
            .status(422)
            .json({ error: 'Project is not eligible for OIDC credential exchange' });
        const issuer = await Issuer.discover(disoveryUrl);

        if (!issuer.metadata.jwks_uri)
          return res
            .status(422)
            .json({ error: 'Project is not eligible for JWKS backed OIDC credential exchange' });
        const jwks = await axios.get(issuer.metadata.jwks_uri);

        if (jwks.status !== 200)
          return res
            .status(422)
            .json({ error: 'Project is not eligible for JWKS backed OIDC credential exchange' });

        let claims = jwt.decode(req.body.token, { complete: true }) as jwt.Jwt | null;
        if (!claims) return res.status(422).json({ error: 'Invalid OIDC token provided' });
        const key = jwks.data.keys.find(key => key.kid === claims!.header.kid);

        if (!key) return res.status(422).json({ error: 'Invalid kid found in the token provided' });

        const pem = jwkToPem(key);
        try {
          claims = jwt.verify(req.body.token, pem, { complete: true }) as jwt.Jwt | null;
        } catch {
          return res
            .status(422)
            .json({ error: 'Could not verify the provided token against the OIDC provider' });
        }

        if (!claims) return res.status(422).json({ error: 'Invalid OIDC token provided' });

        if (
          !(await requester.doOpenIDConnectClaimsMatchProject(
            claims.payload as jwt.JwtPayload,
            project,
            config,
          ))
        ) {
          return res.status(422).json({ error: 'Provided OIDC token does not match project' });
        }

        let githubToken: string;
        try {
          githubToken = await getGitHubAppInstallationToken(project);
        } catch (err) {
          console.error(err);
          return res.status(422).json({ error: 'Failed to obtain access token for project' });
        }

        return res.json({
          GITHUB_TOKEN: githubToken,
        });
      },
    ),
  );

  router.post(
    `/:projectId/${requester.slug}`,
    validate(
      {
        a,
        params: {
          projectId: Joi.number()
            .integer()
            .required(),
        },
      },
      async (req, res) => {
        const project = await getRequestReadyProject(req.params.projectId);
        if (!project) return res.status(404).json({ error: 'Could not find project' });

        const config = requester.getConfigForProject(project);
        if (!config)
          return res.status(422).json({ error: 'Project is not configured to use this requester' });

        const requestMetadata = await requester.metadataForInitialRequest(req, res);
        // Do not send a response here as metadataForInitialRequest will send the response
        if (!requestMetadata) return;

        const newRequest = await OTPRequest.create(
          {
            projectId: project.id,
            state: 'requested',
            requested: new Date(),
            requestMetadata,
            responseMetadata: {},
            proof: OTPRequest.generateProof(),
          },
          {
            returning: true,
          },
        );
        // Fetch with includes
        const completeRequest = (await getFullRequest(`${project.id}`, newRequest.id))!;
        const request = await requester.isOTPRequestValidForRequester(completeRequest);
        if (!request) return res.status(422).json({ error: 'CFA Requester is misconfigured' });

        const allowedState = await requester.validateActiveRequest(request, config);
        if (!allowedState.ok) {
          request.state = 'error';
          request.errored = new Date();
          request.errorReason = allowedState.error;
          await request.save();
          return res.status(400).json({
            error: 'Invalid build provided, check the CFA dashboard for reasoning.',
          });
        }

        res.json(request);
      },
    ),
  );

  router.post(
    `/:projectId/${requester.slug}/:requestId/validate`,
    validate(
      {
        a,
        params: {
          projectId: Joi.number()
            .integer()
            .required(),
          requestId: Joi.string()
            .uuid({ version: 'uuidv4' })
            .required(),
        },
      },
      async (req, res) => {
        const unknownRequest = await getFullRequest(req.params.projectId, req.params.requestId);
        if (!unknownRequest) {
          return res.status(404).json({
            error: 'That request does not exist or is invalid',
          });
        }

        const request = await requester.isOTPRequestValidForRequester(unknownRequest);
        if (!request)
          return res.status(422).json({ error: 'Project is not configured to use this requester' });

        const config = requester.getConfigForProject(request.project);

        if (!config)
          return res
            .status(422)
            .json({ error: 'Project is missing the required configuration to use this requester' });

        if ((config as any).then)
          return res.status(500).json({
            error: 'getConfigForProject returned a promise-like thing and that is not OK',
          });

        if (request.state !== 'requested') {
          return res.status(400).json({
            error: `Expected the request to be in state "requested" but was in state "${request.state}"`,
          });
        }

        const allowedState = await requester.validateActiveRequest(request, config);
        if (!allowedState.ok) {
          request.state = 'error';
          request.errored = new Date();
          request.errorReason = allowedState.error;
          await request.save();
          return res.status(400).json({
            error: 'Invalid build provided, check the CFA dashboard for reasoning.',
          });
        }

        if (!(await requester.validateProofForRequest(request, config))) {
          request.state = 'error';
          request.errored = new Date();
          request.errorReason =
            'Failed to validate build.  Could not find the proof in the build logs in an adaquete time period.';
          await request.save();
          return res.status(403).json({
            error: 'Failed to validate the build, check the CFA dashboard for reasoning.',
          });
        }

        request.state = 'validated';
        request.validated = new Date();
        await request.save();

        // Fetch just the request with no additional info **before** the responder adds its
        // metadata
        const newRequest = (await OTPRequest.findByPk(request.id))!;

        await getResponderFor(request.project).requestOtp(
          request,
          await requester.getRequestInformationToPassOn(request),
        );

        res.json(newRequest);
      },
    ),
  );

  router.post(
    `/:projectId/${requester.slug}/:requestId`,
    validate(
      {
        a,
        params: {
          projectId: Joi.number()
            .integer()
            .required(),
          requestId: Joi.string()
            .uuid({ version: 'uuidv4' })
            .required(),
        },
      },
      async (req, res) => {
        const request = await OTPRequest.findOne({
          where: {
            id: req.params.requestId,
            projectId: req.params.projectId,
          },
        });
        if (!request)
          return res
            .status(404)
            .json({ error: 'That request does not exist or you do not have permission to see it' });
        if (request.state !== 'responded')
          return res.status(204).json({ error: 'That request does not have a response yet' });
        res.json(request);
      },
    ),
  );

  return router;
}
