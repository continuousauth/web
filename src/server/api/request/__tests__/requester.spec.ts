import { Router } from 'express';
import * as request from 'supertest';

import { createRequesterRoutes } from '../requester';
import {
  Project,
  OTPRequest,
  CircleCIRequesterConfig,
  SlackResponderConfig,
} from '../../../db/models';
import { temporaryDatabaseForTestScope } from '../../../../__mocks__/db';

const makeMockRequester = () => ({
  slug: 'test',
  getConfigForProject: jest.fn(),
  metadataForInitialRequest: jest.fn(),
  validateActiveRequest: jest.fn(),
  validateProofForRequest: jest.fn(),
  isOTPRequestValidForRequester: jest.fn(),
  getRequestInformationToPassOn: jest.fn(),
});

describe('requester endpoint creator', () => {
  let router: Router;
  let mock: ReturnType<typeof makeMockRequester>;

  beforeEach(async () => {
    mock = makeMockRequester();
    router = createRequesterRoutes(mock);
  });

  temporaryDatabaseForTestScope();

  describe('request auth', () => {
    it('should reject requests with no auth header', async () => {
      const response = await request(router).delete('/123/test');
      expect(response.status).toBe(403);
      expect(response.text).toMatchInlineSnapshot(
        `"{\\"error\\":\\"Missing Authorization header\\"}"`,
      );
    });

    it('should reject requests for projects that do not exist', async () => {
      const response = await request(router)
        .delete('/123/test')
        .auth('abc', { type: 'bearer' });
      expect(response.status).toBe(404);
      expect(response.text).toMatchInlineSnapshot(
        `"{\\"error\\":\\"Project not found, disabled or you are not authorizad to access it\\"}"`,
      );
    });

    it('should reject requests with an incorrect auth header (non-bearer)', async () => {
      const project = new Project({
        id: 123,
        repoName: 'cfa',
        repoOwner: 'electron',
        secret: 'very scret thing',
        defaultBranch: 'main',
      });
      await project.save();
      const response = await request(router)
        .delete('/123/test')
        .auth('username', 'password');
      expect(response.status).toBe(404);
      expect(response.text).toMatchInlineSnapshot(
        `"{\\"error\\":\\"Project not found, disabled or you are not authorizad to access it\\"}"`,
      );
    });

    it('should reject requests with an incorrect bearer token', async () => {
      const project = new Project({
        id: 123,
        repoName: 'cfa',
        repoOwner: 'electron',
        secret: 'very scret thing',
        defaultBranch: 'main',
      });
      await project.save();
      const response = await request(router)
        .delete('/123/test')
        .auth('the wrong secret thing', { type: 'bearer' });
      expect(response.status).toBe(404);
      expect(response.text).toMatchInlineSnapshot(
        `"{\\"error\\":\\"Project not found, disabled or you are not authorizad to access it\\"}"`,
      );
    });

    it('should reject requests for a project that has not been completely configured', async () => {
      const project = new Project({
        id: 123,
        repoName: 'cfa',
        repoOwner: 'electron',
        secret: 'very scret thing',
        defaultBranch: 'main',
      });
      await project.save();
      const response = await request(router)
        .delete('/123/test')
        .auth('very scret thing', { type: 'bearer' });
      expect(response.status).toBe(400);
      expect(response.text).toMatchInlineSnapshot(
        `"{\\"error\\":\\"Project not completely configured\\"}"`,
      );
    });

    it('should passthrough authed requests for a project that has been configured', async () => {
      const project = new Project({
        id: 123,
        repoName: 'cfa',
        repoOwner: 'electron',
        secret: 'very scret thing',
        defaultBranch: 'main',
      });
      await project.save();
      const circleConfig = await CircleCIRequesterConfig.create(
        {
          accessToken: '123',
        },
        {
          returning: true,
        },
      );
      // await circleConfig.save({ returning: true });
      project.requester_circleCI_id = circleConfig.id;
      const slackConfig = await SlackResponderConfig.create(
        {
          teamName: 'my team',
          teamId: 'my team id',
          teamIcon: 'team icon',
          channelName: 'my channel',
          channelId: 'channel id',
          enterpriseId: 'enterprise id',
          usernameToMention: '@me',
        },
        {
          returning: true,
        },
      );
      project.responder_slack_id = slackConfig.id;
      await project.save();
      const response = await request(router)
        .delete('/123/test')
        .auth('very scret thing', { type: 'bearer' });
      expect(response.status).toBe(404);
      expect(response.text).toMatchInlineSnapshot(`
                                                                                "<!DOCTYPE html>
                                                                                <html lang=\\"en\\">
                                                                                <head>
                                                                                <meta charset=\\"utf-8\\">
                                                                                <title>Error</title>
                                                                                </head>
                                                                                <body>
                                                                                <pre>Cannot DELETE /123/test</pre>
                                                                                </body>
                                                                                </html>
                                                                                "
                                                            `);
    });
  });

  describe('with request auth bypassed', () => {
    beforeEach(() => {
      (global as any).__bypassRequesterAuthentication = true;
    });

    afterEach(() => {
      delete (global as any).__bypassRequesterAuthentication;
    });

    describe('creating a request', () => {
      it('should 404 for an incorrect slug', async () => {
        const response = await request(router).post('/123/voodoo');
        expect(response.status).toEqual(404);
        expect(response.text).toMatchInlineSnapshot(`
                                                                                                                                                                                                                    "<!DOCTYPE html>
                                                                                                                                                                                                                    <html lang=\\"en\\">
                                                                                                                                                                                                                    <head>
                                                                                                                                                                                                                    <meta charset=\\"utf-8\\">
                                                                                                                                                                                                                    <title>Error</title>
                                                                                                                                                                                                                    </head>
                                                                                                                                                                                                                    <body>
                                                                                                                                                                                                                    <pre>Cannot POST /123/voodoo</pre>
                                                                                                                                                                                                                    </body>
                                                                                                                                                                                                                    </html>
                                                                                                                                                                                                                    "
                                                                                                                                                                    `);
      });

      it('should 400 for an invalid project ID', async () => {
        const response = await request(router).post('/bad-id/test');
        expect(response.status).toEqual(400);
        expect(response.text).toMatchInlineSnapshot(
          `"{\\"error\\":\\"Params Validation Error\\",\\"message\\":\\"child \\\\\\"projectId\\\\\\" fails because [\\\\\\"projectId\\\\\\" must be a number]\\"}"`,
        );
      });

      it('should 404 for a project that does not exist', async () => {
        const response = await request(router).post('/123/test');
        expect(response.status).toEqual(404);
        expect(response.text).toMatchInlineSnapshot(`"{\\"error\\":\\"Could not find project\\"}"`);
      });

      describe('with an existing project', () => {
        beforeEach(async () => {
          const project = new Project({
            id: 123,
            repoName: 'cfa',
            repoOwner: 'electron',
            secret: 'very scret thing',
            defaultBranch: 'main',
          });
          await project.save();
        });

        it('should return 422 for a project that is not configured to use this requester', async () => {
          const response = await request(router).post('/123/test');
          expect(response.status).toEqual(422);
          expect(response.text).toMatchInlineSnapshot(
            `"{\\"error\\":\\"Project is not configured to use this requester\\"}"`,
          );
        });

        it('should return the correct code for a project that is configured but fails to acquire metadata', async () => {
          mock.getConfigForProject.mockImplementationOnce(async () => ({
            myConfig: true,
          }));
          mock.metadataForInitialRequest.mockImplementationOnce(async (req, res) => {
            res.status(321).send('metadata error');
            return null;
          });
          const response = await request(router).post('/123/test');
          expect(response.status).toEqual(321);
          expect(response.text).toBe('metadata error');
        });

        it('should create an OTPRequest for a project that is configured and has valid metadata', async () => {
          expect(await OTPRequest.findAll()).toHaveLength(0);
          mock.getConfigForProject.mockImplementationOnce(async () => ({
            myConfig: true,
          }));
          mock.metadataForInitialRequest.mockImplementationOnce(async (req, res) => ({
            metadata: 321,
          }));
          await request(router).post('/123/test');
          const requests = await OTPRequest.findAll();
          expect(requests).toHaveLength(1);
          const { id, proof, requested, ...strippedRequest } = requests[0].get() as OTPRequest;
          expect(strippedRequest).toMatchSnapshot();
        });

        it('should return a 422 if the created OTPRequest is not immediately deemed valid', async () => {
          mock.getConfigForProject.mockImplementationOnce(async () => ({
            myConfig: true,
          }));
          mock.metadataForInitialRequest.mockImplementationOnce(async (req, res) => ({
            metadata: 321,
          }));
          const response = await request(router).post('/123/test');
          expect(response.status).toBe(422);
          expect(response.text).toMatchInlineSnapshot(
            `"{\\"error\\":\\"CFA Requester is misconfigured\\"}"`,
          );
        });

        it('should return a 400 and mark the OTPRequest as errored if the request is not validated', async () => {
          mock.getConfigForProject.mockImplementationOnce(async () => ({
            myConfig: true,
          }));
          mock.metadataForInitialRequest.mockImplementationOnce(async (req, res) => ({
            metadata: 321,
          }));
          mock.isOTPRequestValidForRequester.mockImplementationOnce(request => request);
          mock.validateActiveRequest.mockImplementationOnce(async () => ({
            ok: false,
            error: "that isn't a bird, that's a plane",
          }));
          const response = await request(router).post('/123/test');
          expect(response.status).toBe(400);
          expect(response.text).toMatchInlineSnapshot(
            `"{\\"error\\":\\"Invalid build provided, check the CFA dashboard for reasoning.\\"}"`,
          );
          const requests = await OTPRequest.findAll();
          expect(requests).toHaveLength(1);
          expect(requests[0].state).toBe('error');
          expect(requests[0].errored).toBeTruthy();
          expect(requests[0].errorReason).toMatchInlineSnapshot(
            `"that isn't a bird, that's a plane"`,
          );
        });

        it('should return a 200 and leave the OTPRequest as requested if the request is validated', async () => {
          mock.getConfigForProject.mockImplementationOnce(async () => ({
            myConfig: true,
          }));
          mock.metadataForInitialRequest.mockImplementationOnce(async (req, res) => ({
            metadata: 321,
          }));
          mock.isOTPRequestValidForRequester.mockImplementationOnce(request => request);
          mock.validateActiveRequest.mockImplementationOnce(async () => ({
            ok: true,
          }));
          const response = await request(router).post('/123/test');
          expect(response.status).toBe(200);
          const body = JSON.parse(response.text);
          expect(body).toHaveProperty('id');
          expect(body).toHaveProperty('proof');
          const requests = await OTPRequest.findAll();
          expect(requests).toHaveLength(1);
          expect(requests[0].id).toBe(body.id);
          expect(requests[0].proof).toBe(body.proof);
        });

        it('should generate unique proof strings for consecutive requests', async () => {
          mock.getConfigForProject.mockImplementation(async () => ({
            myConfig: true,
          }));
          mock.metadataForInitialRequest.mockImplementation(async (req, res) => ({
            metadata: 321,
          }));
          mock.isOTPRequestValidForRequester.mockImplementation(request => request);
          mock.validateActiveRequest.mockImplementation(async () => ({
            ok: true,
          }));
          const response = await request(router).post('/123/test');
          expect(response.status).toBe(200);
          const body = JSON.parse(response.text);
          const response2 = await request(router).post('/123/test');
          expect(response2.status).toBe(200);
          const body2 = JSON.parse(response2.text);
          expect(body.proof).not.toBe(body2.proof);
        });
      });
    });

    describe('validating a request', () => {
      const testUuid = '76e9d11d-517a-4fa0-a954-44e3cc6f9a96';

      beforeEach(async () => {
        const project = new Project({
          id: 123,
          repoName: 'cfa',
          repoOwner: 'electron',
          secret: 'very scret thing',
          defaultBranch: 'main',
        });
        await project.save();
      });

      it('should 400 for a bad UUID', async () => {
        const response = await request(router).post('/123/test/not-a-uuid/validate');
        expect(response.status).toBe(400);
        expect(response.text).toMatchInlineSnapshot(
          `"{\\"error\\":\\"Params Validation Error\\",\\"message\\":\\"child \\\\\\"requestId\\\\\\" fails because [\\\\\\"requestId\\\\\\" must be a valid GUID]\\"}"`,
        );
      });

      it('should 400 for a bad project ID', async () => {
        const response = await request(router).post(`/bad-project/test/${testUuid}/validate`);
        expect(response.status).toBe(400);
        expect(response.text).toMatchInlineSnapshot(
          `"{\\"error\\":\\"Params Validation Error\\",\\"message\\":\\"child \\\\\\"projectId\\\\\\" fails because [\\\\\\"projectId\\\\\\" must be a number]\\"}"`,
        );
      });

      it('should 404 for a non-existent request', async () => {
        const response = await request(router).post(`/123/test/${testUuid}/validate`);
        expect(response.status).toBe(404);
        expect(response.text).toMatchInlineSnapshot(
          `"{\\"error\\":\\"That request does not exist or is invalid\\"}"`,
        );
      });

      describe('with an existing request', () => {
        let req: OTPRequest;
        beforeEach(async () => {
          req = new OTPRequest({
            id: testUuid,
            projectId: '123',
            state: 'requested',
            requested: new Date(),
            requestMetadata: {},
            responseMetadata: {},
            proof: 'very-strong-and-secure',
          });
          await req.save();
        });

        it('should 404 for a request for a different project ID', async () => {
          const response = await request(router).post(`/321/test/${testUuid}/validate`);
          expect(response.status).toBe(404);
          expect(response.text).toMatchInlineSnapshot(
            `"{\\"error\\":\\"That request does not exist or is invalid\\"}"`,
          );
        });

        it('should 422 for a request associated with a project that does not have a matching configured requester for the requestmetadata', async () => {
          mock.isOTPRequestValidForRequester.mockImplementation(async () => false);
          const response = await request(router).post(`/123/test/${testUuid}/validate`);
          expect(mock.isOTPRequestValidForRequester).toBeCalled();
          expect(response.status).toBe(422);
          expect(response.text).toMatchInlineSnapshot(
            `"{\\"error\\":\\"Project is not configured to use this requester\\"}"`,
          );
        });

        it('should 422 for a request associated with a project that does not have a valid config for this requester', async () => {
          mock.isOTPRequestValidForRequester.mockImplementation(async r => r);
          mock.getConfigForProject.mockImplementation(() => null);
          const response = await request(router).post(`/123/test/${testUuid}/validate`);
          expect(mock.getConfigForProject).toBeCalled();
          expect(response.status).toBe(422);
          expect(response.text).toMatchInlineSnapshot(
            `"{\\"error\\":\\"Project is missing the required configuration to use this requester\\"}"`,
          );
        });

        it('should 500 for a request associated with a project has a mis-implementeded getConfigForProject as async', async () => {
          mock.isOTPRequestValidForRequester.mockImplementation(async r => r);
          mock.getConfigForProject.mockImplementation(async () => null);
          const response = await request(router).post(`/123/test/${testUuid}/validate`);
          expect(mock.getConfigForProject).toBeCalled();
          expect(response.status).toBe(500);
          expect(response.text).toMatchInlineSnapshot(
            `"{\\"error\\":\\"getConfigForProject returned a promise-like thing and that is not OK\\"}"`,
          );
        });

        it('should 400 for a request that has errored', async () => {
          req.state = 'error';
          await req.save();
          mock.isOTPRequestValidForRequester.mockImplementation(async r => r);
          mock.getConfigForProject.mockImplementation(() => ({ config: 'stuff' }));
          const response = await request(router).post(`/123/test/${testUuid}/validate`);
          expect(response.status).toBe(400);
          expect(response.text).toMatchInlineSnapshot(
            `"{\\"error\\":\\"Expected the request to be in state \\\\\\"requested\\\\\\" but was in state \\\\\\"error\\\\\\"\\"}"`,
          );
        });

        it('should 400 for a request that has been responded to already', async () => {
          req.state = 'responded';
          await req.save();
          mock.isOTPRequestValidForRequester.mockImplementation(async r => r);
          mock.getConfigForProject.mockImplementation(() => ({ config: 'stuff' }));
          const response = await request(router).post(`/123/test/${testUuid}/validate`);
          expect(response.status).toBe(400);
          expect(response.text).toMatchInlineSnapshot(
            `"{\\"error\\":\\"Expected the request to be in state \\\\\\"requested\\\\\\" but was in state \\\\\\"responded\\\\\\"\\"}"`,
          );
        });

        it('should 400 for a request that has been validated already', async () => {
          req.state = 'validated';
          await req.save();
          mock.isOTPRequestValidForRequester.mockImplementation(async r => r);
          mock.getConfigForProject.mockImplementation(() => ({ config: 'stuff' }));
          const response = await request(router).post(`/123/test/${testUuid}/validate`);
          expect(response.status).toBe(400);
          expect(response.text).toMatchInlineSnapshot(
            `"{\\"error\\":\\"Expected the request to be in state \\\\\\"requested\\\\\\" but was in state \\\\\\"validated\\\\\\"\\"}"`,
          );
        });
      });
    });

    describe('getting a validated request', () => {
      const testUuid = '993fb356-c7e0-4a07-9c2a-6e7a37940445';

      beforeEach(async () => {
        const project = new Project({
          id: 123,
          repoName: 'cfa',
          repoOwner: 'electron',
          secret: 'very scret thing',
          defaultBranch: 'main',
        });
        await project.save();
      });

      it('should 400 for a bad UUID', async () => {
        const response = await request(router).post('/123/test/this-is-a-uuid-i-swear');
        expect(response.status).toBe(400);
        expect(response.text).toMatchInlineSnapshot(
          `"{\\"error\\":\\"Params Validation Error\\",\\"message\\":\\"child \\\\\\"requestId\\\\\\" fails because [\\\\\\"requestId\\\\\\" must be a valid GUID]\\"}"`,
        );
      });

      it('should 400 for a bad projectId', async () => {
        const response = await request(router).post(`/bad-id/test/${testUuid}`);
        expect(response.status).toBe(400);
        expect(response.text).toMatchInlineSnapshot(
          `"{\\"error\\":\\"Params Validation Error\\",\\"message\\":\\"child \\\\\\"projectId\\\\\\" fails because [\\\\\\"projectId\\\\\\" must be a number]\\"}"`,
        );
      });

      it('should 404 for a non-existent request', async () => {
        const response = await request(router).post(`/123/test/${testUuid}`);
        expect(response.status).toBe(404);
        expect(response.text).toMatchInlineSnapshot(
          `"{\\"error\\":\\"That request does not exist or you do not have permission to see it\\"}"`,
        );
      });

      it('should 404 for a request that exists but is not for the current project', async () => {
        const otpRequest = new OTPRequest({
          id: testUuid,
          projectId: '123',
          state: 'requested',
          requested: new Date(),
          requestMetadata: {},
          responseMetadata: {},
          proof: 'very-strong-and-secure',
        });
        await otpRequest.save();
        const response = await request(router).post(`/111/test/${testUuid}`);
        expect(response.status).toBe(404);
        expect(response.text).toMatchInlineSnapshot(
          `"{\\"error\\":\\"That request does not exist or you do not have permission to see it\\"}"`,
        );
      });

      it('should 204 No-Content for a request that has not been responded to', async () => {
        const otpRequest = new OTPRequest({
          id: testUuid,
          projectId: '123',
          state: 'requested',
          requested: new Date(),
          requestMetadata: {},
          responseMetadata: {},
          proof: 'very-strong-and-secure',
        });
        await otpRequest.save();
        const response = await request(router).post(`/123/test/${testUuid}`);
        expect(response.status).toBe(204);
        expect(response.text).toMatchInlineSnapshot(`""`);
      });

      it('should return an otp request that has been responded to', async () => {
        const otpRequest = new OTPRequest({
          id: testUuid,
          projectId: '123',
          state: 'responded',
          response: 'my-otp',
          requested: new Date(),
          requestMetadata: {},
          responseMetadata: {},
          proof: 'very-strong-and-secure',
        });
        await otpRequest.save();
        const response = await request(router).post(`/123/test/${testUuid}`);
        expect(response.status).toBe(200);
        const body = JSON.parse(response.text);
        expect(body.id).toBe(otpRequest.id);
        expect(body.response).toBe('my-otp');
      });
    });
  });
});
