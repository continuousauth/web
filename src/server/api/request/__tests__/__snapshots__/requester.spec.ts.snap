// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requester endpoint creator with request auth bypassed creating a request with an existing project should create an OTPRequest for a project that is configured and has valid metadata 1`] = `
{
  "errorReason": null,
  "errored": null,
  "projectId": 123,
  "requestMetadata": {
    "metadata": 321,
  },
  "responded": null,
  "response": null,
  "responseMetadata": {},
  "state": "requested",
  "userThatResponded": null,
  "validated": null,
}
`;
